src/test/java/org/jenkinsci/plugins/workflow/job/WorkflowRunWithFlowExecutionListenerTest.java

```diff
@@ -0,0 +1,114 @@
+import org.jvnet.hudson.test.TestExtension;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import java.io.IOException;
+import java.io.PrintStream;
+
+import static org.junit.Assert.assertNotNull;
+
+public class WorkflowRunWithFlowExecutionListenerTest {
+
+    @ClassRule
+    public static BuildWatcher buildWatcher = new BuildWatcher();
+
+    @Rule
+    public RestartableJenkinsRule story = new RestartableJenkinsRule();
+
+    @Test public void testOnCompleteIsExecutedBeforeListenerIsClosed() throws Exception {
+        story.then(r -> {
+            WorkflowJob p = r.jenkins.createProject(WorkflowJob.class, "p");
+            p.setDefinition(new CpsFlowDefinition("echo 'Running for listener'", true));
+            WorkflowRun b = p.scheduleBuild2(0).waitForStart();
+
+            r.waitForCompletion(b);
+            WorkflowRunWithFlowExecutionListenerTest.Listener listener = ExtensionList.lookup(FlowExecutionListener.class).get(WorkflowRunWithFlowExecutionListenerTest.Listener.class);
+            assertNotNull(listener);
+
+            Thread.sleep( 500);
+            r.assertLogContains("Running", b);
+            r.assertLogContains("blah blah blah", b);
+            r.assertLogContains("Build Number :1", b);
+        });
+    }
+
+
+    @TestExtension("testOnCompleteIsExecutedBeforeListenerIsClosed")
+    public static class Listener extends FlowExecutionListener {
+        @Override
+        public void onCompleted(@Nonnull FlowExecution execution) {
+            super.onCompleted(execution);
+            Queue.Executable executable = null;
+            try {
+                PrintStream logger = execution.getOwner().getListener().getLogger();
+                logger.println("blah blah blah");
+                executable = execution.getOwner().getExecutable();
+                if( executable instanceof WorkflowRun){
+                    WorkflowRun run = (WorkflowRun) executable;
+                    Action action = new Action() {
+                        @CheckForNull
+                        @Override
+                        public String getIconFileName() {
+                            return null;
+                        }
+
+                        @CheckForNull
+                        @Override
+                        public String getDisplayName() {
+                            return null;
+                        }
+
+                        @CheckForNull
+                        @Override
+                        public String getUrlName() {
+                            return null;
+                        }
+                    };
+                    run.addAction(action);
+                    WorkflowJob workflowJob = run.getParent();
+                    int number = workflowJob.getLastBuild().number;
+                    logger.println("Build Number :" + number);
+                }
+            } catch (IOException e) {
+                e.printStackTrace();
```

Comment:
I would rethrow this as an unchecked exception instead so that the build and test would fail if there was a problem:

```suggestion
                throw new RuntimeException(e);
```